tsne/bh_sne_src/tsne.h

```diff
@@ -49,12 +49,37 @@
 #endif
 
 extern "C" {
-
+// stateless t-SNE
 void DLL_PUBLIC run(double* X, int N, int D, double* Y, int no_dims,
                     double perplexity, double theta, int rand_seed,
                     bool skip_random_init, double* init, bool use_init,
                     int max_iter = 1000, int stop_lying_iter = 250,
                     int mom_switch_iter = 250);
+// stateful t-SNE
+struct TSNEState;
+struct TSNE {
+  // parameters
+  int N;
+  double* Y;
+  int no_dims;
+  double theta;
+  int max_iter;
+  int stop_lying_iter;
+  int mom_switch_iter;
+  struct TSNEState* state;
+  int iter;
+  float total_time;
+  float residual_time;
+};
```

Comment:
This should also be a forward-declare, and the actual struct definition should be outside of the `extern "C"` block so that it can use `unique_ptr<TSNEState>` rather than `TSNEState *`.  `TSNEState` is already not a `C` type because it's full of `vector<double>`.  Or even better (and more complex implementation-wise) `vector<double>` for `Y`.  Or just make `state` not be a pointer.  I don't really see why it's a pointer.

Also you should probably reorder a few of the fields for better alignment and packing.  `float` and `int` are 32-bit fields so should be ordered together separately from the pointer fields which have to be 64-bit aligned, and the `double` fields which perform better when 64-bit aligned.

Reply:
I made `state` a pointer so that it could be forward-declared and `TSNE` could be a C type (`TSNEState` doesn't have to be, it's opaque to C).

Reply:
But as far as the C interface is concerned `TSNE` is an opaque pointer too.  Types are not part of the the C abi, after all.  Where it gets relevant is for cases where you're not just using the ABI, but also interrogating the header file to learn something about the API, e.g. cgo or rust bindgen.  In that case, I think it would be reasonable to stick `state` at the end of the struct and basically telling such consumers to ignore it as something they don't need to pay attention to - lots of structs defined in C APIs have a "`reserved`" region which presumably the internal code understands but is supposed to be opaque to the API consumers.

Reply:
Alternatively, treat `TSNE` as really-opaque as in "you aren't supposed to even know how big it is" and provide accessor methods for the public fields.  Which might be a better idea anyway to ensure users don't violate an invariants.